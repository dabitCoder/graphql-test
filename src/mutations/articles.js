import { gql } from "@apollo/client";

export const CREATE_ARTICLE = gql`
  mutation CreateArticle($title: String!, $url: String!, $language: String!) {
    createArticle(title: $title, url: $url, language: $language) {
      title
      url
      language
    }
  }
`;
