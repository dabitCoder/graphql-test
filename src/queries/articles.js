import { gql } from "@apollo/client";

export const GET_ARTICLES = gql`
  query getArticles($page: Int, $pageSize: Int) {
    paginatedArticles(page: $page, pageSize: $pageSize) {
      entries {
        id
        title
        url
        language
      }
      meta {
        pageNumber
        pageSize
      }
    }
  }
`;
