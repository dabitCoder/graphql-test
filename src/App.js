import React, { useState } from "react";
import { useQuery } from "@apollo/client";
import { Spinner } from "@chakra-ui/react";

import { GET_ARTICLES } from "./queries/articles";
import Table from "./Table";
import Form from "./Form";

function App() {
  const [currentPage, setCurrentPage] = useState(1);

  const { loading, data, fetchMore } = useQuery(GET_ARTICLES, {
    variables: { page: currentPage, pageSize: 5 },
  });

  const handleLoadMore = () => {
    setCurrentPage(currentPage + 1);
    fetchMore({
      variables: {
        page: currentPage,
      },
    });
  };

  const renderArticles = () => {
    if (data) {
      return <Table data={data} />;
    }
  };

  if (loading) {
    return (
      <Spinner
        thickness="4px"
        speed="0.65s"
        emptyColor="gray.200"
        color="blue.500"
        size="xl"
      />
    );
  } else {
    return (
      <Form handleLoadMore={handleLoadMore} renderArticles={renderArticles} />
    );
  }
}

export default App;
