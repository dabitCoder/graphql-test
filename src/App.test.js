import React from "react";
import {
  render,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import App from "./App";
import { MockedProvider } from "@apollo/react-testing";
import { GET_ARTICLES } from "./queries/articles";

const defaultMock = [
  {
    request: {
      query: GET_ARTICLES,
    },
    result: {
      data: {
        allArticles: [{ id: 1, title: "Test title", url: "test.url.com" }],
      },
    },
  },
];

describe("App testing", () => {
  const component = (mocks = defaultMock) => (
    <MockedProvider mocks={mocks} addTypename={false}>
      <App />
    </MockedProvider>
  );

  const setup = (mocks) => render(component(mocks));

  it("Should match snapshot", async () => {
    const { container } = setup();
    await waitForElementToBeRemoved(screen.getByText("Loading..."));
    expect(container).toMatchSnapshot();
  });

  it("Should render loading", () => {
    setup();
    const loadingText = screen.getByText("Loading...");
    expect(loadingText).toBeDefined();
  });

  it("Should render an input", async () => {
    setup();
    const titleInput = await screen.findByLabelText("article-title");
    expect(titleInput).toBeDefined();
  });
});
