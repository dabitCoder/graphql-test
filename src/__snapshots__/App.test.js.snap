// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App testing Should match snapshot 1`] = `
<div>
  <div
    class="css-466ycc"
  >
    <div
      class="css-gmuwbf"
    >
      <p
        class="chakra-text css-1gjuw9l"
      >
        Create a new article
      </p>
    </div>
    <div
      class="chakra-stack css-17kdbol"
    >
      <input
        aria-label="article-title"
        class="chakra-input css-0"
        placeholder="Article Title"
      />
      <input
        class="chakra-input css-0"
        placeholder="Article url"
      />
      <input
        class="chakra-input css-0"
        placeholder="Article Language"
      />
      <button
        aria-label="new-article"
        class="chakra-button css-5c2176"
        type="button"
      >
        Create article
      </button>
    </div>
    <div
      class="css-gmuwbf"
    >
      <p
        class="chakra-text css-1gjuw9l"
      >
        Current articles
      </p>
    </div>
    <table
      class="chakra-table css-0"
      role="table"
    >
      <thead
        class="css-0"
      >
        <tr
          class="css-0"
          role="row"
        >
          <th
            class="css-0"
          >
            Article title
          </th>
          <th
            class="css-0"
          >
            Article url
          </th>
        </tr>
      </thead>
      <tbody
        class="css-0"
      >
        <tr
          class="css-0"
          role="row"
        >
          <td
            class="css-0"
            role="gridcell"
          >
            Test title
          </td>
          <td
            class="css-0"
            role="gridcell"
          >
            test.url.com
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
